pallets/otc-settlements/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+		///
+		/// The profit made by closing the arbitrage is transferred to `FeeReceiver`.
+		///
+		/// Parameters:
+		/// - `origin`: Signed or unsigned origin. Unsigned origin doesn't pay the TX fee,
+		/// 			but can be submitted only by a collator.
+		/// - `otc_id`: ID of the OTC order with existing arbitrage opportunity.
+		/// - `amount`: Amount necessary to clone the arb.
+		/// - `route`: The route we trade against. Required for the fee calculation.
+		///
+		/// Emits `Executed` event when successful.
+		///
+		#[pallet::call_index(0)]
+		#[pallet::weight(<T as Config>::WeightInfo::settle_otc_order()
+			.saturating_add(<T as Config>::RouterWeightInfo::sell_weight(route))
+		.saturating_add(<T as Config>::RouterWeightInfo::get_route_weight())
+		.saturating_add(<T as Config>::RouterWeightInfo::calculate_spot_price_with_fee_weight(route))
+		.saturating_add(<T as pallet_otc::Config>::WeightInfo::fill_order().max(<T as pallet_otc::Config>::WeightInfo::partial_fill_order()))
+		)]
+		pub fn settle_otc_order(
+			_origin: OriginFor<T>,
+			otc_id: OrderId,
+			amount: Balance,
+			route: Vec<Trade<AssetIdOf<T>>>,
+		) -> DispatchResult {
+			Self::settle_otc(otc_id, amount, route)
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	pub fn account_id() -> T::AccountId {
+		PALLET_ID.into_account_truncating()
+	}
+
+	/// Ensure that the profit is more than some minimum amount.
+	fn ensure_min_profit(otc_amount_in: Balance, _profit: Balance) -> DispatchResult {
+		let _min_expected_profit = T::MinProfitPercentage::get().mul_floor(otc_amount_in);
+
+		// In the benchmark we calculate the overhead of extrinsic and we doesn't make any trade.
+		// We disable this check because otherwise it would fail.
+		#[cfg(not(feature = "runtime-benchmarks"))]
+		// tell the binary search algorithm to find higher values
+		ensure!(_profit >= _min_expected_profit, Error::<T>::TradeAmountTooLow);
+
+		Ok(())
+	}
+
+	/// Executes two trades: asset_a -> OTC -> asset_b, and asset_b -> Router -> asset_a.
+	///
+	/// If the OTC order is partially fillable, the extrinsic fails if the existing arbitrage
+	/// opportunity is not closed after the trade and if there is no profit after the trade.
+	/// If the OTC order is not partially fillable, fails only if there is no profit after the trade.
+	pub fn settle_otc(otc_id: OrderId, amount: Balance, route: Vec<Trade<AssetIdOf<T>>>) -> DispatchResult {
+		let pallet_acc = Self::account_id();
+
+		let otc = <pallet_otc::Orders<T>>::get(otc_id).ok_or(Error::<T>::OrderNotFound)?;
+		let (asset_a, asset_b) = (otc.asset_in, otc.asset_out);
+
+		// get initial account balances
+		let asset_a_balance_before = <T as Config>::Currency::free_balance(asset_a, &pallet_acc);
+		let asset_b_balance_before = <T as Config>::Currency::free_balance(asset_b, &pallet_acc);
+
+		<T as Config>::Currency::deposit(asset_a, &pallet_acc, amount)?;
+
+		if !otc.partially_fillable {
+			ensure!(otc.amount_out == amount, Error::<T>::NotPartiallyFillable);
+		}
```

Comment:
you can move this above, so no need to query the free and balances and make deposit.